src/main/scala/mesosphere/marathon/api/HTTPMetricsFilter.scala

```diff
@@ -0,0 +1,89 @@
+package mesosphere.marathon
+package api
+
+import javax.servlet._
+import javax.servlet.http.{ HttpServletRequest, HttpServletRequestWrapper, HttpServletResponse, HttpServletResponseWrapper }
+import mesosphere.marathon.metrics.{ Metrics, ServiceMetric }
+
+/**
+  * This filter replaces the default I/O streams with proxies that count
+  * the number of bytes that are sent to and from the client.
+  */
+class HTTPMetricsFilter extends Filter {
+  private[this] val inputBytesMetric = Metrics.counter(ServiceMetric, getClass, "inputBytes")
+  private[this] val outputBytesMetric = Metrics.counter(ServiceMetric, getClass, "outputBytes")
+
+  /**
+    * Wraps a `ServletOutputStream` and overrides the `write` method in
+    * order to count the number of bytes that went through the stream.
+    *
+    * @param upstream The stream to proxy the output to
+    */
+  private class ProxyOutputStream(val upstream: ServletOutputStream) extends ServletOutputStream {
+    var bytes: Int = 0
+
+    override def isReady: Boolean = upstream.isReady
+    override def setWriteListener(writeListener: WriteListener): Unit = upstream.setWriteListener(writeListener)
+    override def write(b: Int): Unit = {
```

Comment:
Are you sure that bytes are read/written one by one? What about overriding implementations of other read/write methods like `read(byte[] b)`, `read(byte[] b, int off, int len)`, `readLine(byte[] b, int off, int len)` or so on?

Reply:
Well, the `read(): Int` and `write(b: Int)` are the only abstract methods of the base stream classes that **must** be implemented. Therefore I assumed that the default implementation of the other signatures is implemented by calling out to these base functions. And I confirmed this by checking the counters while polling various HTTP endpoints.

However, now that I think of it, the stream implementations _might_ provide a more efficient implementation for operations on byte ranges. So it actually makes sense to implement the other signatures of `read` and `write`. (But `readLine` does not really benefit from this.)

Reply:
My understanding matches your suspicions, @wavesoft : all other methods call out to write(int).